rig-core/src/providers/mira.rs

```diff
@@ -0,0 +1,439 @@
+//! Mira API client and Rig integration
+//!
+//! # Example
+//! ```
+//! use rig::providers::mira;
+//!
+//! let client = mira::Client::new("YOUR_API_KEY");
+//!
+//! ```
+use crate::{
+    agent::AgentBuilder,
+    completion::{self, CompletionError, CompletionRequest},
+    extractor::ExtractorBuilder,
+    message::{self, Message},
+    OneOrMany,
+};
+use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION, CONTENT_TYPE};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::string::FromUtf8Error;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum MiraError {
+    #[error("Invalid API key")]
+    InvalidApiKey,
+    #[error("API error: {0}")]
+    ApiError(u16),
+    #[error("Request error: {0}")]
+    RequestError(#[from] reqwest::Error),
+    #[error("UTF-8 error: {0}")]
+    Utf8Error(#[from] FromUtf8Error),
+}
+
+#[derive(Debug, Deserialize)]
+struct ApiErrorResponse {
+    message: String,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+enum ApiResponse<T> {
+    Ok(T),
+    Err,
+}
+
+#[derive(Debug, Serialize)]
+pub struct AiRequest {
+    pub model: String,
+    pub messages: Vec<ChatMessage>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub temperature: Option<f32>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub max_tokens: Option<u32>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub stream: Option<bool>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ChatMessage {
+    pub role: String,
+    pub content: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct CompletionResponse {
+    pub choices: Vec<ChatChoice>,
+    pub usage: Option<Usage>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ChatChoice {
+    pub message: ChatMessage,
+}
+
+#[derive(Debug, Deserialize)]
+struct ModelsResponse {
+    data: Vec<ModelInfo>,
+}
+
+#[derive(Debug, Deserialize)]
+struct ModelInfo {
+    id: String,
+}
+
+#[derive(Clone)]
+/// Client for interacting with the Mira API
+pub struct Client {
+    base_url: String,
+    client: reqwest::Client,
+    headers: HeaderMap,
+}
+
+impl Client {
+    /// Create a new Mira client with the given API key
+    pub fn new(api_key: impl AsRef<str>) -> Result<Self, MiraError> {
```

Comment:
We generally just do `api_key: &str`, is there a reason why you used `impl AsRef<str>`?